### ISSUE TITLE
[proposal] Provide a way to retrieve the data of all textfields within a `Form` widget

### ISSUE THREAD

**Issue by casvanluijtelaar** [flutter/flutter#107528]
## Use case

the Flutter `Form` widget can be extremely powerful and has a lot of potential to simplify form implementations. Sadly I'm finding myself having to create my own Form widgets because there is not simple build-in way to aggregate all the form data in one place. Just like the form can interact with all its `FormField` children through validation, saving, ... it should also have an option to submit it's childrens data. Similar to how HTML handles form submission.

## Proposal

since the `FormState` already has all its `FormFieldState` children registered it would be trivial to collect all their values. the implementation difficulties come from returning the form data in a manageable way.

I would propose adding a `submissionKey` string to the `FormField<T>` base class. this allows us to aggregate all the `FormField` data under a `Map<String, dynamic>` where the entries would be shaped as: `submissionKey:  value`. 

this would require all `FormField` to have a non-null submissionKey. This can be hande by either throwing an exception when calling a form submit on a null submissionKey, or not adding `FormField` values to the end `Form` submit data if the submissionKey is null.

This would be a minor addition to the framework with no breaking or changing code, that would still have enormous benefits.

comparing the current closest implementation of this:
```dart
Form(
  child: Column(
    children: [
      TextField( 
        onSave: (value) {
          data['name'] = value;
        },
      ),
      TextField(
        onSave: (value) {
          data['email'] = value;
        },
      ),
      TextField(
         onSave: (value) {
          data['message'] = value;
        },
      ),
      MaterialButton(
        child: Text('submit'),
        onPressed: () {
          final form = Form.of(context);
          if(!form.validate()) return;

          form.save();
          /// handle data in stateful widget...
        }
      ),
    ],
  ),    
);
```

would be turned into something like this: 
```dart
Form(
  child: Column(
    children: [
      TextField(
        submissionKey: 'name',
        ...
      ),
      TextField(
        submissionKey: 'email',
        ...
      ),
      TextField(
        submissionKey: 'message',
        ...
      ),
      MaterialButton(
        child: Text('submit'),
        onPressed: () {
          final form = Form.of(context);
          if(!form.validate()) return;
          
          final data = form.submit();
          // final object = FormObject.fromJson(data); ...
        }
      );
    ],
  ),    
);

```

**Comment by pedromassango** [comment]
So this ticket is about adding a way to retrieve the data of all fields?

**Comment by casvanluijtelaar** [comment]
> So this ticket is about adding a way to retrieve the data of all fields?

yes, All FormField descendants of this Form widget with a valid submissionKey

**Comment by fngadiyo** [comment]
2 years later, still no plan for this.

**Comment by Mairramer** [comment]
I’m implementing something that addresses the proposal described in
[https://github.com/flutter/flutter/issues/110760](https://github.com/flutter/flutter/issues/110760)
and also **partially** covers a related proposal.

The main idea is to allow extracting the values of all `FormField`s from the `FormState` in a structured way, using `ValueKey`s as identifiers.

Example:

```dart
final formKey = GlobalKey<FormState>();

Scaffold(
  body: Form(
    key: formKey,
    child: Column(
      children: <Widget>[
        TextFormField(
          key: const ValueKey('name'),
          initialValue: 'Name',
        ),
        TextFormField(
          key: const ValueKey('email'),
          initialValue: 'Email',
        ),
      ],
    ),
  ),
);

final List<FormFieldState<dynamic>> fields =
    formKey.currentState!.fields.toList();

final data = <String, dynamic>{};

for (final field in fields) {
  final key = field.widget.key;
  if (key is ValueKey<String>) {
    data[key.value] = field.value;
  }
}

print(data);
// {name: Name, email: Email}
```

This makes it possible to obtain a `Map<String, dynamic>` representing the current state of the form, without having to keep individual references to each field.
